src/React.Router/HtmlHelperExtensions.cs

```diff
@@ -0,0 +1,167 @@
+		/// <param name="htmlTag">HTML tag to wrap the component in. Defaults to &lt;div&gt;</param>
+		/// <param name="containerId">ID to use for the container HTML tag. Defaults to an auto-generated ID</param>
+		/// <param name="clientOnly">Skip rendering server-side and only output client-side initialisation code. Defaults to <c>false</c></param>
+		/// <param name="serverOnly">Skip rendering React specific data-attributes during server side rendering. Defaults to <c>false</c></param>
+		/// <param name="containerClass">HTML class(es) to set on the container tag</param>
+		/// <returns><see cref="IHtmlString"/> containing the rendered markup for provided React Router component</returns>
+		public static IHtmlString ReactRouterWithContext<T>(
+			this IHtmlHelper htmlHelper, 
+			string componentName,
+			T props,
+			string path = null,
+			HttpResponse Response = null,
+			string htmlTag = null,
+			string containerId = null,
+			bool clientOnly = false,
+			bool serverOnly = false,
+			string containerClass = null,
+			Action<HttpResponse, RoutingContext> contextHandler = null
+		)
+		{
+			try
+			{
+				path = path ?? htmlHelper.ViewContext.HttpContext.Request.Path;
+				Response = Response ?? htmlHelper.ViewContext.HttpContext.Response;
+
+				var reactComponent 
+					= Environment.CreateRouterComponent
+						(componentName, props, path, containerId, clientOnly);
+
+				if (!string.IsNullOrEmpty(htmlTag))
+				{
+					reactComponent.ContainerTag = htmlTag;
+				}
+				if (!string.IsNullOrEmpty(containerClass))
+				{
+					reactComponent.ContainerClass = containerClass;
+				}
+				
+				var executionResult = reactComponent.RenderRouterWithContext(clientOnly, serverOnly);
+
+				if (executionResult.Context?.status != null)
+				{
+					// Use provided contextHandler
+					if (contextHandler != null)
+					{
+						contextHandler(Response, executionResult.Context);
+					}
+					// Handle routing context internally
+					else
+					{
+						HandleRoutingContext(executionResult.Context, Response);
+					}
+				}
+
+				return new HtmlString(executionResult.RenderResult);
+			}
+			finally
+			{
+				Environment.ReturnEngineToPool();
+			}
+		}
+
+		private static void HandleRoutingContext(RoutingContext context, HttpResponse Response)
+		{
+			var statusCode = context.status.Value;
+
+			// 300-399
+			if (statusCode >= 300 && statusCode < 400)
+			{
+				if (!string.IsNullOrEmpty(context.url))
+				{
+					if (statusCode == 301)
+					{
+
+#if NET451
+
+						Response.RedirectPermanent(context.url);
+#else
+						Response.Redirect(context.url, true);
+#endif
+					}
+					else // 302 and all others
+					{
+						Response.Redirect(context.url);
+					}
+				}
+				else
+				{
+					throw new ReactRouterException("Router requested redirect but no url provided.");
+				}
+			}
+			else
+			{
+				Response.StatusCode = statusCode;
```

Comment:
It feels a bit strange for a HTML helper to set the status code, I can't think of a better way to do this though.